schnorr_fun/src/share_backup.rs

```diff
@@ -0,0 +1,238 @@
+    polynomial: Vec<Point>,
+) -> [u5; 4] {
+    let hash = H::default();
+    hash.add(&polynomial[..]).finalize().to_base32()[0..4]
+        .try_into()
+        .expect("4 bech32 chars must fit 4 character arry")
+}
+
+/// Create a bech32m secret share backup
+///
+/// Requires that the threshold be no greater than 1024.
+/// If using an integer (small) index, the backup will be 67 bech32 characters.
+/// If using a scalar index, the backup will be 118 bech32 characters.
+pub fn encode_backup<H: Default + Digest<OutputSize = U32>>(
+    polynomial: &Vec<Point>,
+    secret_share: &Scalar,
+    share_index: &Scalar<Public>,
+) -> String {
+    if poly::point::eval(polynomial, *share_index) != g!(secret_share * G) {
+        panic!("Secret share is not valid with respect to the polynomial!")
+    }
+
+    let threshold = polynomial.len();
+    let mut data = [u5::default(); 2 + 4 + 52 + 52];
+
+    if threshold > 1024 {
+        panic!("Polynomial has too high of a threshold, {threshold} > 1024");
+    }
+    let threshold_u5 = (threshold - 1).to_le_bytes().to_vec().to_base32();
+    for (i, byte) in threshold_u5.into_iter().take(2).enumerate() {
+        data[i] = byte;
+    }
+
+    let polynomial_identifier_u5 = polynomial_identifier::<H>(polynomial.clone());
+    for (i, byte) in polynomial_identifier_u5.into_iter().enumerate() {
+        data[2 + i] = byte;
+    }
+
+    let secret_share_u5 = secret_share.to_bytes().to_vec().to_base32();
+    for (i, byte) in secret_share_u5.into_iter().enumerate() {
+        data[2 + 4 + i] = byte;
+    }
+
+    let is_small =
+        share_index.to_bytes()[0..31].iter().all(|b| *b == 0) && share_index.to_bytes()[31] < 32;
+
+    let n_unused_bytes = if is_small {
+        let share_index_u5 =
+            u5::try_from_u8(share_index.to_bytes()[31]).expect("must be less than 32");
+        data[2 + 4 + 52] = share_index_u5;
+        52 - 1
+    } else {
+        let share_index_u5 = share_index.to_bytes().to_base32();
+        for (i, byte) in share_index_u5.iter().enumerate() {
+            data[2 + 4 + 52 + i] = *byte;
+        }
+        0
+    };
+
+    bech32::encode("frost", &data[..(data.len() - n_unused_bytes)], Bech32m)
+        .expect("hrp must be valid")
+}
+
+/// Decode a bech32m secret share backup
+pub fn decode_backup(
+    encoded: String,
+) -> Result<(usize, [u5; 4], Scalar, Scalar), FrostBackupDecodeError> {
+    let (hrp, data, variant) =
+        bech32::decode(&encoded).map_err(FrostBackupDecodeError::Bech32DecodeError)?;
+
+    if hrp != "frost" {
+        return Err(FrostBackupDecodeError::InvalidHumanReadablePrefix);
+    }
+
+    if !matches!(variant, bech32::Variant::Bech32m) {
+        return Err(FrostBackupDecodeError::WrongBech32Variant(variant));
+    }
+
+    let mut threshold_bytes =
+        Vec::<u8>::from_base32(&data[..2]).map_err(FrostBackupDecodeError::Bech32DecodeError)?;
+    threshold_bytes.resize((usize::BITS / 8) as usize, 0);
```

Comment:
I had to do this to work on 32bit architectures, reason enough to use something instead of usize for threshold?

Reply:
This should just be a u16